### DISCUSSION TITLE
assembly anisotropic diffusion

### DISCUSSION THREAD

**Discussion by marCahana** [firedrakeproject/firedrake#3050] [Firedrake support]
Hi all, 

I would like to solve anisotropic diffusion of the following form: 

$- \nabla \ \cdot \ (C \nabla u) = f, \ \text{in} \  \Omega$
$\qquad \qquad \quad u=0, \ \text{on} \  \partial \Omega,$
where $C \in R^{d \times d}$, where $d$ is dimension of the problem. 

I created simple example: 
```
        ncells = 10
        mesh = UnitSquareMesh(ncells, ncells, quadrilateral=True, comm=COMM_WORLD, reorder=True)
        V    = FunctionSpace(mesh ,"CG", 1)

        bc         = DirichletBC(V, 0.0, [1, 2, 3, 4]) 

        v_test     = TestFunction(V)

        f_func = Function(V)
        f_func.dat.data[:] = sampled_rhs[:]

        sol = Function(V)

        coef = 1e-5
        D = as_tensor(((1, 0), (0, coef)))

        F = inner(D*grad(sol), grad(v_test))*dx - f_func*v_test*dx

        problem = NonlinearVariationalProblem(F, sol, bcs=bc)
        solver  = NonlinearVariationalSolver(problem)
        solver.solve()

```

I am guessing this approach has some problem as I am getting the following warning messages:
```
form0_cell_integral_otherwise: check array access within bounds: completed (0.38s wall 1.00x CPU)
wrap_form0_cell_integral_otherwise: check array access within bounds: completed (0.46s wall 1.00x CPU)
```

In cases, where I run the code multiple times for varying right-hand sides and coefficients, the code crashes stating that memory is out of bounds. 

Would anybody have a recommendation in order to fix the problem?

**Comment by ksagiyam** [comment]
I cannot reproduce the error, but I saw:
```
isl/isl_ctx.c:295: isl_ctx freed, but some objects still reference it
```
when I ran your example hot-cache.

With the firedrake venv active, could you try:
```
pip uninstall islpy
firedrake-update
```
and see if this resolves the issue?

**Reply by marCahana** [reply]
Many thanks for your help. 
I tried your suggestion, but I am still getting the same warnings and eventually the errors. 

Is there any way to clean up cache or memory related to code generation during the code execution? 
Let's say, in between running the above code for different sets of coefficients/right-hand-sides.

**Reply by ksagiyam** [reply]
`firedrake-clean` cleans the cache, but it is probably not safe to call it from within a python script if parallel.

We might have to see if someone could reproduce the issue. Could you hard-code `sampled_rhs` (probably with a smaller `ncells`) and provide us with a deterministic way to reproduce the issue? Does, e.g.,:
```
firedrake-clean
python your_script.py
python your_script.py
```
always make your code crush?
